### PR TITLE
LIS-5351: Add initial GitHub actions build to cut a release

### PR DESCRIPTION
**Pull request by blakesmith**
Triggers on a new tag being pushed.

1. Runs the build, and integration test suite.
2. Signs artifacts, and pushes to maven central.
3. Creates a new GitHub release.

### CODE REVIEW

**Review by erik-helleren — DISMISSED** [review]

File: .github/workflows/release-and-publish-artifacts.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      - '*'
+jobs:
+  publish:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Setup build dependencies
+        run: make setup
+      - name: Set up Java for publishing to Maven Central Repository
+        uses: actions/setup-java@v1
```
Comment (david-huber) [inline .github/workflows/release-and-publish-artifacts.yml:14]: why not use `@v2`?

Comment (blakesmith) [reply .github/workflows/release-and-publish-artifacts.yml:14]: Didn't know there was v2.

File: .github/workflows/release-and-publish-artifacts.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        with:
+          java-version: 1.8
```
Comment (david-huber) [inline .github/workflows/release-and-publish-artifacts.yml:16]: Is this okay? Docs on the action indicate `8` is an acceptable value, but not the old school `1.8` version format.

Comment (blakesmith) [reply .github/workflows/release-and-publish-artifacts.yml:16]: Will update.

File: .github/workflows/release-and-publish-artifacts.yml [hunk 3]
```diff
@@ -0,0 +1,33 @@
+name: release and push to central
+on:
+  push:
+    tags:
+      - '*'
```
[the remark this answers is not in the document]
Comment (david-huber) [reply .github/workflows/release-and-publish-artifacts.yml:5]: Correct, only members of various sproutsocial teams have write access to this repository. creating a tag ref cannot be done by a non-member.

File: .github/workflows/release-and-publish-artifacts.yml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+          java-version: 1.8
+          server-id: ossrh
+          server-username: MAVEN_USERNAME
+          server-password: MAVEN_PASSWORD
+          gpg-private-key: ${{ secrets.OSSRH_GPG_SECRET_KEY }}
+          gpg-passphrase: MAVEN_GPG_PASSPHRASE
+      - name: Publish to the Maven Central Repository
+        run: make clean deploy
+        env:
+          MAVEN_USERNAME: ${{ secrets.OSSRH_USERNAME }}
```
Comment (david-huber) [inline .github/workflows/release-and-publish-artifacts.yml:25]: do these need to be envvars, or can they be passed directly in the `with:` yaml stanza? if so please move them there.

File: .github/workflows/release-and-publish-artifacts.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (blakesmith) [reply .github/workflows/release-and-publish-artifacts.yml:5]: Good catch. I believe I can narrow this further, and will do so:

```yaml
on:
  push:
    branches:    
      - main
    # Sequence of patterns matched against refs/tags
    tags:        
      - v*
```

Docs here: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-including-branches-and-tags

[the remark this answers is not in the document]
Comment (balanceiskey) [reply .github/workflows/release-and-publish-artifacts.yml:5]: Tag pushes themselves should be locked down as write access on the repo is limited to Sprout org members.

File: .github/workflows/release-and-publish-artifacts.yml [hunk 5]
```diff
@@ -0,0 +1,33 @@
+name: release and push to central
+on:
+  push:
+    tags:
+      - '*'
+jobs:
+  publish:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
Comment (balanceiskey) [inline .github/workflows/release-and-publish-artifacts.yml:10]: Similar question to below, any opposition to using @v3 here? -- https://github.com/actions/checkout/releases

File: .github/workflows/release-and-publish-artifacts.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (david-huber) [reply .github/workflows/release-and-publish-artifacts.yml:5]: I'm not strictly opposed to the change you're making right now @blakesmith, but let's say we discover a critical security issue two major versions back. A good practice would be to branch that major version, apply the fix, then tag the branch to fix the issue.

File: .github/workflows/release-and-publish-artifacts.yml (hunk 2 above)
Comment (david-huber) [reply .github/workflows/release-and-publish-artifacts.yml:16]: See also https://github.com/actions/setup-java/blob/main/docs/switching-to-v2.md

File: .github/workflows/release-and-publish-artifacts.yml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+          gpg-private-key: ${{ secrets.OSSRH_GPG_SECRET_KEY }}
+          gpg-passphrase: MAVEN_GPG_PASSPHRASE
+      - name: Publish to the Maven Central Repository
+        run: make clean deploy
+        env:
+          MAVEN_USERNAME: ${{ secrets.OSSRH_USERNAME }}
+          MAVEN_PASSWORD: ${{ secrets.OSSRH_TOKEN }}
+          MAVEN_GPG_PASSPHRASE: ${{ secrets.OSSRH_GPG_SECRET_KEY_PASSWORD }}
+      - name: Create release
+        uses: ncipollo/release-action@v1
```
Comment (balanceiskey) [inline .github/workflows/release-and-publish-artifacts.yml:29]: Just a general reminder for ourselves, we'll need to add `ncipollo/release-action@v1` to the allow list at the org level once we bring this in. 

(General note for other reviewers, this action is directly referenced in GitHub's now-archived version of a similar action: https://github.com/actions/create-release)

File: .github/workflows/release-and-publish-artifacts.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (blakesmith) [reply .github/workflows/release-and-publish-artifacts.yml:5]: @balanceiskey Yup, I just realized this as well. I do think we should be explicit about tags that start with `v*` or something still. But yeah, we don't actually want all pushes to main to re-publish to maven central.

@david-huber I'm not sure what you're advocating for. Are you ok with any tag pushed to `v*` publishing to maven central?

**Review by balanceiskey — CHANGES_REQUESTED** [review]
Some minor stuff, but lookin' good 👍

File: .github/workflows/release-and-publish-artifacts.yml (hunk 6 above)
Comment (blakesmith) [reply .github/workflows/release-and-publish-artifacts.yml:29]: Yup, I already took care of this. :-)

File: .github/workflows/release-and-publish-artifacts.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (blakesmith) [reply .github/workflows/release-and-publish-artifacts.yml]: Update to only push on tags that match `v*`, let me know if you have a different idea here!

File: .github/workflows/release-and-publish-artifacts.yml (hunk 2 above)
Comment (blakesmith) [reply .github/workflows/release-and-publish-artifacts.yml]: Looks like v1 supported the legacy version syntax, while v2 does not. This should be fixed.

**Comment by blakesmith** [comment]
@balanceiskey Can you take another look whenever you get the chance?

**Review by balanceiskey — APPROVED** [review]

**Pull request merged**